arbor/gpu_context.cpp

```diff
@@ -17,23 +32,43 @@ bool gpu_context::has_atomic_double() const {
     return attributes_ & gpu_flags::has_atomic_double;
 }
 
+bool gpu_context::has_gpu() const {
+    return id_ != -1;
+}
+
 #ifndef ARB_HAVE_GPU
 
-gpu_context::gpu_context(): has_gpu_(false), attributes_(0) {}
 void gpu_context::synchronize_for_managed_access() const {}
+gpu_context::gpu_context(int) {
+    throw arbor_exception("Arbor must be compiled with CUDA support to select a GPU.");
+}
 
 #else
 
-gpu_context::gpu_context(): has_gpu_(true), attributes_(0) {
+gpu_context::gpu_context(int gpu_id) {
     cudaDeviceProp prop;
-    cudaGetDeviceProperties(&prop, 0);
+    auto status = cudaGetDeviceProperties(&prop, gpu_id);
+    if (status==cudaErrorInvalidDevice) {
+        throw arbor_exception("Invalid GPU id " + std::to_string(gpu_id));
```

Comment:
The hierarchy was mainly there to keep our messages and error contextual data consistent. My feeling is that one-offs in `arbor_exeception` are okay, but there is certainly scope for consistently wrapping this one too, or cuda-error code wrapping exceptions, or so on.

Reply:
What Sam said. This was a one-off, though it might make sense to see whether there are other CUDA-related exceptions, and collate them. But that is beyond the scope of this PR.